Cookbook/Proposals/FeatureSwitchesRedesign.md

```diff
@@ -0,0 +1,218 @@
+
+### App configuration boilerplate
+
+We use app configuration type (and types related to it) to define "static" feature switches things that are hardcoded in the app and are cuased by different requirements for the products. I.e. one app can define that it should use a new feature flow while another does not use it.
+
+This brings more clarity in what values are defined for what apps, we don't have to consult Firebase config and we don't risk accidently changing the value. We also don't need to wait for firebase config to be fetched and don't risk not having a right value when Firebase fails. 
+
+At the same time it makes us to repeat all the configurations in each target which is a lot of boilerplate for a simple flag and these configs grew a lot over time. Or we have to use protocol extensions with default values which makes it trickier to understand the actual value for particular target as we need to make sure this default is not overriden anywhere else.
+
+### Local Feature Switches boilerplate
+
+It takes a lot of boilerplate to define a local feature switch - we need to change the code in several places (in one file though), need to manually add an entry to the `Root.plist` and make sure that it's in the right place and in the write file. We need to make sure we move it to the different section (for remote feature toggles) if we move the flag to the Firebase, otherwise we won't be able to override it in the Settings app and it will always have a default value if we override any other flag.
+
+
+### Debugging
+
+It's not clear what feature switches affect what screens. It's somtimes frustrating to go through a complex flow and then realise that the right feature switch was not turned on/off in the Settings app or local overrides were active and to go through the flow again (as we don't observe changes in the feature switches at runtime and require application to restart).
+
+
+## Proposed solution
+
+To address these issues we suggest the following changes in our feature switches architecture:
+
+1. Feature toggles and static configurations related to the particular feature should be defined in this feature module instead of shared module or in the app. This follows interface segregation principle and goes along with improving modularisation - one won't need to change common module with feature switches and recompile all of it (and potentially everything else) when they only need to recompile module that uses this feature switch. This way we also configure our frameworks the same way as we configure any other 3rd party framework and can use the similar infrastrauctore for that (i.e. dedicated instances responsibule for configuring each feature rather than doing that all in the app delegate).
```

Comment:
🥇 I keep thinking of centralisation but this is a very valid approach, with the main benefit being quicker iteration and less recompiling.

I'm not sure about discoverability though (but that should be enforced with code generation).